core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -91,9 +91,9 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
     case IO.Tags.Fail =>
       val io = self.asInstanceOf[IO.Fail[E]]
 
-      new IO.Fail(f(io.error), io.defects)
+      new IO.Fail(io.cause.map(f))
 
-    case _ => new IO.Redeem(self, (e: E) => new IO.Fail(f(e), Nil), (a: A) => new IO.Strict(g(a)))
+    case _ => new IO.Redeem(self, (cause: Cause[E]) => IO.fail0(cause.map(f)), (a: A) => new IO.Strict(g(a)))
```

Comment:
`bimap` can be expressed as `leftMap(f).map(g)`, then we don't have to worry if this logic is correct.

Reply:
Also you'll delete like 15 lines of code. 😄